### ISSUE TITLE
Points, vectors, coordinates, and charts/atlases

### ISSUE THREAD

**Issue by sethaxen** [JuliaManifolds/Manifolds.jl#6]
This is an issue for discussing how the concepts of points, vectors, coordinates and charts should be represented/implemented.

In my conception, points are abstract and do not assume a specific coordinate system. I don't see how these can be represented. What we have been calling point is more like coordinates, which assumes a chart. See https://github.com/JuliaNLSolvers/ManifoldProjections.jl/issues/6#issuecomment-499390236.  Tangent vectors are only defined wrt a specific point.

A major decision discussed in several places is whether to define specific types for coordinates and vectors. I personally like this idea, since we can potentially dispatch on the type of coordinates/vector and prevent nonsensical operations like adding vectors from different tangent spaces. Perhaps points/vectors should be subtypes of `AbstractArray` that store an `AbstractArray` with the actual values. And how would we dispatch based on point identity?

For matrix groups, perhaps we should have tangent matrices as well.

**Comment by mateuszbaran** [comment]
In my opinion the basic interface should be independent of a specific representation of points, tangent vectors, coordinates etc. I think this is possible. Specific manifolds can be implemented in different ways with different trade-offs. I would be OK with multiple implementations of the same manifold.

**Comment by kellertuer** [comment]
As far as I think and as I started with the `MPoint` and `TVector` types – if you just don't care and use `AbstractArrays` its fine, the `ArrayManifold` enhanced these with tests, but you could also take a manifold now, different subtypes of `MPoint` and `TVector` and dispatch on these, so that one manifold could be implemented with different charts or coordinates by specifying point and tangent vector types. With this, the basic interface (on arrays) is still independent, but if you wish to, you can introduce that for your manifold. and coordinate changes are then `convert`s maybe.

**Comment by sdaxen** [comment]
Oh, I think @kellertuer's last comment here is where I got the idea that `MPoint` and `TVector` could be used to dispatch on different representations, even if those are also vectors and matrices. See discussion at https://github.com/JuliaNLSolvers/ManifoldMuseum/pull/18#issuecomment-505192259.

**Comment by kellertuer** [comment]
Yes, that‘s exactly the idea of the `ArrayManifold`, to enhance even vector and matrix manifolds with such additional knowledge, whether it’s a point or a vector. With the convert it even provides that automatically, which I hope is a good way for dispatch: if you don’t care, be lazy and it works (i Hope it just does).

**Comment by kellertuer** [comment]
Reading this after quite some time – of course all our points are currently representations of the points on manifolds, for example using the embedding for the sphere. And we have the generic types for points (and vectors).

So I think one way to tackle the issue here would be a `ChartManifold` or `AtlasManifold` that describes a manifold using an atlas of charts (maps from the manifold to some Rn) or its inverses (parametrizations). One can locally then talk about coordinates wrt. a chart. This would all be far away from being Riemannian.

I think a `MetricManifold{AtlasManifold}` could be done, even with AD of the parametrizations to have a basis for the TpM with respect to which one can specify the metric, but one would need a way to define this. 

The questions remains: What would be the goal of this generic approach? In most cases that I am working with I hope for a more closed-form-approach to a manifold than charts and computing geodesics by ODEs.

Do we have an application for this we could discuss a modelling of the approach with?

**Comment by mateuszbaran** [comment]
I don't have any applications for charts. Retractions and inverse retractions are enough for my needs. I think we can just leave this issue open and work on other parts of the library.

**Comment by kellertuer** [comment]
Me neither, and yes for me retractions are also fine enough; but sure we can leave this open to discuss how one could do this; I just noticed that we are missing an application (which might make it easier to discuss how to implement this feature).

**Comment by mateuszbaran** [comment]
I thought about it a little and maybe having support for charts would be one way to do sampling of point from manifolds? This way we could have bijections between a neighborhood of a point on a manifold and R^n, something retractions and inverse retractions don't provide.

The interface could be very simple, something like
```julia
function get_coordinates(M::Manifold, A::AbstractAtlas, p, q) end
```
for transforming point `q` on manifold `M` to coordinates from the chart from atlas `A` at point (or, in principle, any indexing method we need) `p` to R^n, and 
```julia
function get_point(M::Manifold, A::AbstractAtlas, p, x) end
```
to do the reverse. Maybe we could also add another function for transition maps in case there could be a better implementation than the naive `get_point` and then `get_coordinates`.

This would also let us fix all the metric functions that actually depend on a chart but currently don't, for example multilinear operators need to have coordinates in a chart (https://github.com/JuliaManifolds/Manifolds.jl/pull/202).

Ref. https://github.com/TuringLang/Bijectors.jl/issues/58, https://github.com/JuliaManifolds/Manifolds.jl/pull/28, https://github.com/JuliaManifolds/Manifolds.jl/issues/57, https://discourse.julialang.org/t/spaces-for-base-measures/50594

**Comment by kellertuer** [comment]
This looks neat! Can we do that maybe as a `AbstractChatManifold` as well as a `ChartManifold{Manifold,AbstractAtlas}` ? that acts similar to a `MetricManifold` and we would keep it as just one parameter in the `get_coordinates` function?

This would make it possible to still post different metrics on such a Manifold, though the type is then a little obstructive
`MetricManifold{ChartManifold{Manifold,AbstractAtlas}}`.

Despite this remark I think the approach is great! It would enable to model surfaces for example, though internally we are facing the probing of ho to determine the proper chart (is the point in the domain? Which one do we choose, where `p`s in the domain? These questions).

**Comment by mateuszbaran** [comment]
> Can we do that maybe as a `AbstractChatManifold` as well as a `ChartManifold{Manifold,AbstractAtlas}` ? that acts similar to a `MetricManifold` and we would keep it as just one parameter in the `get_coordinates` function?

Good question. It would be more complex to set up I think, I certainly don't look forward to adding another decorator but I can see some benefits.

> Despite this remark I think the approach is great! It would enable to model surfaces for example, though internally we are facing the probing of ho to determine the proper chart (is the point in the domain? Which one do we choose, where `p`s in the domain? These questions).

Thanks! Chart selection I think is application-dependent. For example for distributions you would likely want a chart as flat as possible near the mean of that distribution (if it exists). If you wanted to solve the ODE for the geodesic, it would be a different choice. If someone was making an actual atlas of the Earth, charts could be indexed by page numbers for example, or names of places.

**Comment by kellertuer** [comment]
Sure, as a decorator it would be more complicated to fit the inerhitance.

Sure, chart selection is very application dependent; I just came up with that due to the `get_coordinates` function, which should have a strategy how to select the chart as a parameter or something like that.

**Comment by mateuszbaran** [comment]
> Sure, chart selection is very application dependent; I just came up with that due to the `get_coordinates` function, which should have a strategy how to select the chart as a parameter or something like that.

The parameter `p` is for selection of the chart. Or did you mean something else here?

**Comment by kellertuer** [comment]
Ah, I misread that, sure; `p` is the thing I meant, I think it could even be a strategy, not only an index (index of a chart I assume?).
Not that the index of a chart might be any indexing type (on the sphere we could center a projection based chart at any point go the sphere). But a generic p to determine the chart is a good idea, yes.

**Comment by mateuszbaran** [comment]
Yes, `p` would be an index of a chart in an atlas.

**Comment by mateuszbaran** [comment]
> I think it could even be a strategy

How would that work?

**Comment by kellertuer** [comment]
Index could also be a real value or vector (for example if you take the atlas of projections onto tangent planes on the sphere, the charts would be indexed by all points p on the sphere).

Strategy in the sense of:
* The „best fitting chart“ for example in the atlas above the one with the closest center/projection point)
* The first where q is in the domain, if you iterate  
* ...

I don't have that many strategies in mind yet but since q might lie in many charts, there has to be the possibility to provide a strategy of choosing a chart to work in.

**Comment by mateuszbaran** [comment]
OK, right, I see.

**Comment by mateuszbaran** [comment]
Solved in #325
